libs/wingc/src/jsify/codemaker.rs

```diff
@@ -1,36 +1,125 @@
 /// `{` and `}`?
-#[derive(Default)]
+#[derive(Default, Clone, Debug)]
 pub struct CodeMaker {
-	lines: Vec<(usize, String)>,
-	indent: usize,
+	lines: Vec<LineData>,
+	indent: IndentAmount,
+	source: Option<WingSpan>,
+}
+
+pub type IndentAmount = usize;
+pub type CharacterOffset = usize;
+
+#[derive(Default, Clone, Debug)]
+pub struct LineData {
+	pub line: String,
+	pub indent: IndentAmount,
+	pub mappings: Vec<(CharacterOffset, WingSpan)>,
+}
+
+/// Converts stringable data into a codemaker
+/// This handles splitting newlines
+fn string_to_code<S: Into<String>>(s: S) -> CodeMaker {
+	let text: String = s.into();
+	let mut code = CodeMaker::default();
+
+	// remove trailing newline
+	let line_text = text.strip_suffix("\n").unwrap_or(&text);
+
+	// if the line has newlines in it, consider each line separately
+	for subline in line_text.split('\n') {
+		code.push_line(LineData {
+			line: subline.to_string(),
+			indent: 0,
+			mappings: vec![],
+		});
+	}
+
+	code
+}
+
+impl Into<CodeMaker> for String {
+	fn into(self) -> CodeMaker {
+		string_to_code(self)
+	}
+}
+
+impl Into<CodeMaker> for &String {
+	fn into(self) -> CodeMaker {
+		string_to_code(self)
+	}
+}
+
+impl Into<CodeMaker> for &str {
+	fn into(self) -> CodeMaker {
+		string_to_code(self)
+	}
+}
+
+impl Into<CodeMaker> for Vec<CodeMaker> {
+	fn into(self) -> CodeMaker {
+		let mut code: CodeMaker = CodeMaker::default();
+		for (idx, line) in self.into_iter().enumerate() {
+			if idx > 0 {
+				code.append(", ");
+			}
+			code.append(line);
+		}
+		code
+	}
 }
 
 impl CodeMaker {
+	pub fn with_source<S: Into<WingSpan>>(source: S) -> Self {
+		Self {
+			source: Some(source.into()),
+			indent: 0,
+			lines: vec![],
+		}
+	}
+
 	/// Emits a line of code and then increases the indent by one.
-	pub fn open<S: Into<String>>(&mut self, line: S) {
+	pub fn open<S: Into<CodeMaker>>(&mut self, line: S) {
 		self.line(line);
 		self.indent += 1;
 	}
 
 	/// Decreases the indent by one and then emits a line of code.
-	pub fn close<S: Into<String>>(&mut self, line: S) {
+	pub fn close<S: Into<CodeMaker>>(&mut self, line: S) {
 		self.indent -= 1;
 		self.line(line);
 	}
 
 	/// Emits a line of code with the current indent.
-	pub fn line<S: Into<String>>(&mut self, line: S) {
-		let line: String = line.into();
+	pub fn line<S: Into<CodeMaker>>(&mut self, line: S) {
+		let mut new_code: CodeMaker = line.into();
 
-		// remove trailing newline
-		let line = line.strip_suffix("\n").unwrap_or(&line);
-
-		// if the line has newlines in it, consider each line separately
-		for subline in line.split('\n') {
-			self.lines.push((self.indent, subline.into()));
+		for mut line_data in new_code.lines.drain(..) {
+			line_data.indent += self.indent;
+			if line_data.mappings.is_empty() {
+				if let Some(source) = self.source.as_ref() {
+					line_data.mappings.push((0, source.clone()))
+				}
+			}
+			self.push_line(line_data);
 		}
```

Comment:
(optional) Another possible way to write this

```suggestion
		for mut line_data in new_code.lines.drain(..) {
            let new_line_data = LineData {
                line: line_data.line,
                indent: line_data.indent + self.indent,
                mappings: line_data.mappings.unwrap_or((0, source.clone())),
            };
			self.push_line(new_line_data);
		}
```

Reply:
Rewrote it to be closer to your suggestion, although because `source` is an Option it had to be a little different